### PR TITLE
Refinement

### PR DESCRIPTION
**Pull request by eed3si9n**
Fixes https://github.com/tree-sitter/tree-sitter-scala/issues/265

Problem
-------
Compound type with refinement (structural type with a parent) doesn't parse.

Solution
--------
This adds support for it.
It collides with context bounds `A: B: C`, so the precedence is set to -1.

### CODE REVIEW

File: script/smoke_test.sh [hunk 1]
```diff
@@ -3,9 +3,9 @@
 # This is an integration test to generally check the quality of parsing.
 
 SCALA_SCALA_LIBRARY_EXPECTED=95
-SCALA_SCALA_COMPILER_EXPECTED=84
+SCALA_SCALA_COMPILER_EXPECTED=87
```
Comment (eed3si9n) [inline script/smoke_test.sh:6]: Scala 2 compiler bumps to `87%` with this change.

File: script/smoke_test.sh [hunk 2]
```diff
@@ -3,9 +3,9 @@
...
 DOTTY_COMPILER_EXPECTED=81
-SYNTAX_COMPLEXITY_CEILING=2300
+SYNTAX_COMPLEXITY_CEILING=2800
```
Comment (ckipp01) [inline script/smoke_test.sh:8]: Do we have an idea of how much is too much here? Like when should it cause refactoring to account for the bump?

**Review by ckipp01 — APPROVED** [review]
Looks good to me! LGTM!

File: script/smoke_test.sh (hunk 2 above)
Comment (eed3si9n) [reply script/smoke_test.sh:8]: I guess when we observe compilation time or memory usage to degrade?

**Pull request merged**